### PR TITLE
Add validating webhook

### LINKED ISSUE #718: [Bug] KubeRay does not show clear error for duplicated `groupName` field
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

### What happened + What you expected to happen

1. Saw: When a cluster YML specifies 2 worker pod types with the same `groupName`, the cluster goes into a loop of creating and immediately terminating all pods with that `groupName`
2. Expected: The autoscaler or `kuberay-operator` fails with a clear error that "Multiple worker pod types have the same `groupName`"

Running Kuberay 0.3.0 and nightly, Ray v2.0.1

### Reproduction script

Use any cluster YML with two worker pod types, both with the same `groupName`

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### LINKED ISSUE #736: [Feature] Better validation for the RayCluster CRD, via validating webhooks
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

There are many aspects of KubeRay config validation that cannot be handled by CRD schema validation.

We have some limited validation functionality within the operator -- if the operator detects certain misconfigurations, it will refuse to reconcile and will instead write and some sort of error state to the status field.
[Code reference.](https://github.com/ray-project/kuberay/blob/3581b91b71dc0ce1bbf33904b4c0cf817a9767a9/ray-operator/controllers/ray/common/pod.go#L811)

However, it would be a better user experience if the CR simply failed validation up submission.
The standard way to do this would be to take advantage of KubeBuilder's support for validating webhooks.

This would make it easy to resolve UX issues like the one described here:
https://github.com/ray-project/kuberay/issues/718

### Use case

Less painful configuration experience.

### Related issues

https://github.com/ray-project/kuberay/issues/718

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by davidxia**
and validate worker group names are unique.

Add new Makefile targets `install-with-webhooks`, `uninstall-with-webhooks`,
`deploy-with-webhooks`, and `undeploy-with-webhooks` to be backwards compatible
and opt-in.

Much of the code, especially the YAML files, was generated by running the
command below as [documented in kubebuilder][1].

```
kubebuilder create webhook \
  --group ray \
  --version v1 \
  --kind RayCluster \
  --defaulting \
  --programmatic-validation
```

## How to use locally

```shell
make manifests generate
make install-with-webhooks
IMG=kuberay/operator:test make docker-build
kind load docker-image kuberay/operator:test
IMG=kuberay/operator:test make deploy-with-webhooks
```

## Example RayCluster that has duplicate worker group names

```shell
cat dupe-worker-group-name.yaml

apiVersion: ray.io/v1
kind: RayCluster
metadata:
  name: dupe-worker-group-name
spec:
  headGroupSpec:
    rayStartParams:
      dashboard-host: '0.0.0.0'
    template:
      spec:
        containers:
        - name: ray-head
          image: rayproject/ray:2.7.0
  workerGroupSpecs:
  - replicas: 1
    minReplicas: 1
    maxReplicas: 10
    groupName: group1
    rayStartParams: {}
    template:
      spec:
        containers:
        - name: ray-worker
          image: rayproject/ray:2.7.0
  - replicas: 1
    minReplicas: 1
    maxReplicas: 10
    groupName: group1
    rayStartParams: {}
    template:
      spec:
        containers:
        - name: ray-worker
          image: rayproject/ray:2.7.0
```

## Before

```
kubectl apply -f dupe-worker-group-name.yaml
raycluster.ray.io/raycluster-dupe-worker-name created
```

## After

```
kubectl --context kind-kind apply -f config/samples/ray-cluster-dupe-worker-name.yaml
```

`The RayCluster "raycluster-dupe-worker-name" is invalid:
spec.workerGroupSpecs[1]: Invalid value: v1.WorkerGroupSpec{GroupName:"group1",
Replicas:(*int32)(0x40006e63cc), MinReplicas:(*int32)(0x40006e63c8),
MaxReplicas:(*int32)(0x40006e63c0), RayStartParams:map[string]string{},
Template:v1.PodTemplateSpec{ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"",
Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0,
CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC),
DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil),
Labels:map[string]string(nil), Annotations:map[string]string(nil),
OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil),
ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)},
Spec:v1.PodSpec{Volumes:[]v1.Volume(nil), InitContainers:[]v1.Container(nil),
Containers:[]v1.Container{v1.Container{Name:"ray-worker",
Image:"rayproject/ray:2.7.0", Command:[]string(nil), Args:[]string(nil),
WorkingDir:"", Ports:[]v1.ContainerPort(nil), EnvFrom:[]v1.EnvFromSource(nil),
Env:[]v1.EnvVar(nil),
Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil),
Requests:v1.ResourceList(nil)}, VolumeMounts:[]v1.VolumeMount(nil),
VolumeDevices:[]v1.VolumeDevice(nil), LivenessProbe:(*v1.Probe)(nil),
ReadinessProbe:(*v1.Probe)(nil), StartupProbe:(*v1.Probe)(nil),
Lifecycle:(*v1.Lifecycle)(nil), TerminationMessagePath:"",
TerminationMessagePolicy:"", ImagePullPolicy:"",
SecurityContext:(*v1.SecurityContext)(nil), Stdin:false, StdinOnce:false,
TTY:false}}, EphemeralContainers:[]v1.EphemeralContainer(nil),
RestartPolicy:"", TerminationGracePeriodSeconds:(*int64)(nil),
ActiveDeadlineSeconds:(*int64)(nil), DNSPolicy:"",
NodeSelector:map[string]string(nil), ServiceAccountName:"",
DeprecatedServiceAccount:"", AutomountServiceAccountToken:(*bool)(nil),
NodeName:"", HostNetwork:false, HostPID:false, HostIPC:false,
ShareProcessNamespace:(*bool)(nil),
SecurityContext:(*v1.PodSecurityContext)(nil),
ImagePullSecrets:[]v1.LocalObjectReference(nil), Hostname:"", Subdomain:"",
Affinity:(*v1.Affinity)(nil), SchedulerName:"",
Tolerations:[]v1.Toleration(nil), HostAliases:[]v1.HostAlias(nil),
PriorityClassName:"", Priority:(*int32)(nil),
DNSConfig:(*v1.PodDNSConfig)(nil), ReadinessGates:[]v1.PodReadinessGate(nil),
RuntimeClassName:(*string)(nil), EnableServiceLinks:(*bool)(nil),
PreemptionPolicy:(*v1.PreemptionPolicy)(nil), Overhead:v1.ResourceList(nil),
TopologySpreadConstraints:[]v1.TopologySpreadConstraint(nil),
SetHostnameAsFQDN:(*bool)(nil), OS:(*v1.PodOS)(nil)}},
ScaleStrategy:v1.ScaleStrategy{WorkersToDelete:[]string(nil)}}: worker group
names must be unique`

## Backwards compatibility

Just use original Makefile targets of `install`, `uninstall`, `deploy`, and
`undeploy`.

```shell
IMG=kuberay/operator:test make docker-build
kind load docker-image kuberay/operator:test
IMG=kuberay/operator:test make deploy

kubectl apply -f dupe-worker-group-name.yaml
raycluster.ray.io/raycluster-dupe-worker-name created
```
closes #718
closes #736

[1]: https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation

## Checks

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/Makefile
Comment (davidxia) [inline ray-operator/Makefile:110]: See https://cert-manager.io/docs/installation/#default-static-install linked from [kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/cert-manager).

File: ray-operator/apis/ray/v1/raycluster_webhook.go
Comment (davidxia) [inline ray-operator/apis/ray/v1/raycluster_webhook.go:81]: important part

File: ray-operator/apis/ray/v1/webhook_suite_test.go
Comment (davidxia) [inline ray-operator/apis/ray/v1/webhook_suite_test.go:166]: relevant test, rest of this file mostly generated by kubebuilder

File: ray-operator/main.go
Comment (davidxia) [inline ray-operator/main.go:176]: copied from kubebuilder docs

- https://book.kubebuilder.io/cronjob-tutorial/running.html?highlight=ENABLE_WEBHOOKS#optional
- https://book.kubebuilder.io/cronjob-tutorial/main-revisited.html?highlight=ENABLE_WEBHOOKS#you-said-something-about-main

File: ray-operator/config/certmanager/certificate.yaml
Comment (davidxia) [inline ray-operator/config/certmanager/certificate.yaml:1]: All the YAML below was either generated by kubebuilder or cargo-culted from a fresh kubebuilder-generated repo that had a lot of original boilerplate for webhooks and certmanager preserved.

**Comment by kevin85421** [comment]
Cool! Can users determine whether to install this webhook or not?

**Comment by davidxia** [comment]
I need to find a way to make it opt-in. Lmk if you have some examples.

File: ray-operator/main.go
Comment (davidxia) [reply ray-operator/main.go:176]: @kevin85421 shall we invert this conditional so that users have to explicitly set the env var to enable the hooks?

Comment (kevin85421) [reply ray-operator/main.go:176]: Yes, disabling it by default is better. Many users are not allowed to install webhooks.

File: ray-operator/Makefile
Comment (davidxia) [inline ray-operator/Makefile:123]: See https://cert-manager.io/docs/installation/#default-static-install linked from [kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/cert-manager).

File: ray-operator/Makefile
Comment (davidxia) [inline ray-operator/Makefile]: created new targets to be backwards compatible and so users can opt-in

File: ray-operator/config/crd-with-webhooks/kustomization.yaml
Comment (davidxia) [inline ray-operator/config/crd-with-webhooks/kustomization.yaml:7]: relative paths so we don't have to duplicate these large files

File: ray-operator/Makefile
Comment (davidxia) [inline ray-operator/Makefile:113]: `--load-restrictor LoadRestrictionsNone` is needed to allow kustomize to load files from outside root of its context. Otherwise, we get the error

```
make uninstall-with-webhooks

test -s /Users/dxia/src/github.com/ray-project/kuberay/ray-operator/bin/controller-gen || GOBIN=/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/bin/controller-gen/.. go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.0
/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/bin/controller-gen "crd:maxDescLen=0,trivialVersions=true,preserveUnknownFields=false,generateEmbeddedObjectMeta=true,allowDangerousTypes=true" rbac:roleName=kuberay-operator webhook paths="./..." output:crd:artifacts:config=config/crd/bases
test -s /Users/dxia/src/github.com/ray-project/kuberay/ray-operator/bin/kustomize ||  GOBIN=/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/bin/kustomize/.. go install sigs.k8s.io/kustomize/kustomize/v5@latest
/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/bin/kustomize build config/crd-with-webhooks | grep -v 'creationTimestamp: "null"' | kubectl delete -f -
Error: accumulating resources: accumulation err='accumulating resources from 'bases/ray.io_rayclusters.yaml': security; file '/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/config/crd-with-webhooks/bases/ray.io_rayclusters.yaml' is not in or below '/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/config/crd-with-webhooks'': must build at directory: '/Users/dxia/src/github.com/ray-project/kuberay/ray-operator/config/crd-with-webhooks/bases/ray.io_rayclusters.yaml': file is not directory
No resources found
make: *** [uninstall-with-webhooks] Error 1
```

File: ray-operator/main.go
Comment (davidxia) [reply ray-operator/main.go]: updated

File: ray-operator/main.go
Comment (davidxia) [inline ray-operator/main.go:176]: adapted from kubebuilder docs

- https://book.kubebuilder.io/cronjob-tutorial/running.html?highlight=ENABLE_WEBHOOKS#optional
- https://book.kubebuilder.io/cronjob-tutorial/main-revisited.html?highlight=ENABLE_WEBHOOKS#you-said-something-about-main

File: ray-operator/config/default-with-webhooks/manager_webhook_patch.yaml
Comment (davidxia) [inline ray-operator/config/default-with-webhooks/manager_webhook_patch.yaml:13]: this env var enables the webhooks

File: ray-operator/Makefile
Comment (davidxia) [inline ray-operator/Makefile:113]: created new targets to be backwards compatible and so users can opt-in

**Comment by davidxia** [comment]
Why does the Go-build-and-test / Lint (pull_request) check fail when it has no files to analyze?

```
run golangci-lint
  Running [/home/runner/golangci-lint-1.54.1-linux-amd64/golangci-lint run --out-format=github-actions --path-prefix=./apiserver --timeout=3m] in [/home/runner/work/kuberay/kuberay/apiserver] ...
  level=error msg="Running error: context loading failed: no go files to analyze"
  
  Error: golangci-lint exit with code 5
  Ran golangci-lint in 6317ms
```

**Comment by kevin85421** [comment]
@davidxia Would you mind fixing the CI errors? Thanks!

**Comment by kevin85421** [comment]
The lint check in CI keeps failing.

**Comment by davidxia** [comment]
> The lint check in CI keeps failing.

should be fixed now!

File: ray-operator/apis/ray/v1/webhook_suite_test.go
Comment (kevin85421) [inline ray-operator/apis/ray/v1/webhook_suite_test.go:1]: Will this test suite be executed by running `make test`?

**Review by kevin85421 — COMMENTED** [review]
Thank you for the contributions! Your contributions are always very high quality. Would you mind resolving the conflicts with the master branch? It looks good to me, but I will also try to test it manually before I merge it.

File: ray-operator/Makefile
Comment (davidxia) [reply ray-operator/Makefile:115]: I copied the pattern from the `install:` target. I can change it, but maybe it's better to keep it consistent?

File: ray-operator/Makefile
Comment (davidxia) [reply ray-operator/Makefile:136]: > Are there any compatibility limitations between cert-manager v1.13.2 and Kubernetes, controller-runtime, or other dependencies?

cert-manager 1.13 is the [latest release and support K8s versions 1.23 — 1.28](https://cert-manager.io/docs/releases/#currently-supported-releases). Does this work?

> Is it necessary to deploy cert-manager if users want to use the webhook?

I think it's possible to run webhooks with insecure communication (something like `insecure-skip-tls-verify`) between K8s control plane and the webhook server, but it's not recommended. I don't even see it mentioned in https://book.kubebuilder.io/. We should probably stick with the guidelines here. wdyt?

File: ray-operator/Makefile
Comment (davidxia) [reply ray-operator/Makefile:140]: see reply to https://github.com/ray-project/kuberay/pull/1584/files#r1402835009

File: ray-operator/apis/ray/v1/raycluster_webhook.go
Comment (davidxia) [reply ray-operator/apis/ray/v1/raycluster_webhook.go:23]: This comment is used for code generation of various K8s YAMLs. See [webhook marker docs](https://book.kubebuilder.io/reference/markers/webhook.html)

* path: specifies that path that the API server should connect to this webhook on
* mutating: marks this as a mutating webhook
* ...

File: ray-operator/apis/ray/v1/raycluster_webhook.go
Comment (davidxia) [reply ray-operator/apis/ray/v1/raycluster_webhook.go:34]: This means this webhook endpoint is for validation instead of mutation.

> marks this as a mutating webhook (it's validating only if false)

— https://book.kubebuilder.io/reference/markers/webhook.html

File: ray-operator/main.go
Comment (davidxia) [reply ray-operator/main.go]: thanks, updated

File: ray-operator/apis/ray/v1/webhook_suite_test.go
Comment (davidxia) [reply ray-operator/apis/ray/v1/webhook_suite_test.go:1]: yup, look for `Running Suite: Webhook Suite`. I also checked by putting `Fail()` in the test and seeing `make test` fails.

File: ray-operator/PROJECT
Comment (davidxia) [reply ray-operator/PROJECT:1]: most of the YAMLs and this were created or modified by (noted in PR description)

```
kubebuilder create webhook \
  --group ray \
  --version v1 \
  --kind RayCluster \
  --defaulting \
  --programmatic-validation
```

File: ray-operator/config/certmanager/certificate.yaml
Comment (davidxia) [reply ray-operator/config/certmanager/certificate.yaml:1]: yes, this page https://book.kubebuilder.io/cronjob-tutorial/running-webhook#deploy-webhooks

**Comment by davidxia** [comment]
Should I update [docs here](https://github.com/ray-project/kuberay/blob/master/docs/deploy/installation.md) in a subsequent PR?

**Comment by kevin85421** [comment]
> Should I update [docs here](https://github.com/ray-project/kuberay/blob/master/docs/deploy/installation.md) in a subsequent PR?

We currently move almost all docs to the Ray website. Maybe we can add a doc for webhook under [User Guides](https://docs.ray.io/en/latest/cluster/kubernetes/user-guides.html) instead?

**Comment by kevin85421** [comment]
Thanks, I will merge it after I manually test it.

**Comment by davidxia** [comment]
Manual tests include

* `make install-with-webhooks && make uninstall-with-webhooks` exits 0
* `make deploy-with-webhooks && make undeploy-with-webhooks` exits 0
* `make install && make install-with-webhooks` exits 0
* `make deploy && make deploy-with-webhooks` exits 0
* `make install-with-webhooks deploy-with-webhooks && kubectl apply -f dupe-worker-group-name.yaml` (see above for file) results in error with message `worker group names must be unique`

Should we write proper tests for these? Is it easy to do so?

**Comment by kevin85421** [comment]
I tested this PR manually by running the following 

```sh
kind create cluster --image=kindest/node:v1.26.0
IMG=kuberay/operator:test make docker-image
kind load docker-image kuberay/operator:test

# If this command reports any errors, rerun it.
IMG=kuberay/operator:test make deploy-with-webhooks

kubectl apply -f dupe-worker-group-name.yaml
# The RayCluster "dupe-worker-group-name" is invalid: spec.workerGroupSpecs[1]: Invalid value: ...... : worker group names must be unique
```

The command `make deploy-with-webhooks` always reports errors about `cert-manager` in the first run. I need to rerun the command when the cert-manager Pods are ready. Is there any way to make the first try succeed?

Screen Shot 2023-11-27 at 12 48 10 PM

**Comment by davidxia** [comment]
Do `serving-cert` and `selfsigned-issuer` take some time to appear on the cluster?

**Comment by kevin85421** [comment]
> Do serving-cert and selfsigned-issuer take some time to appear on the cluster?

No. I do not dig into the root cause, but, in my observation, both `serving-cert` and `selfsigned-issuer` can only be created after cert-manager Pods are ready. If I do not rerun the command, the KubeRay operator Pod will hang forever because ` Warning  FailedMount  11s (x6 over 27s)  kubelet            MountVolume.SetUp failed for volume "cert" : secret "webhook-server-cert" not found`. After I run the command again, the `certificate` will be created immediately.

**Comment by kevin85421** [comment]
Compatibility:

* cert-manager v1.13.2 supports Kubernetes 1.23 - 1.28. (https://github.com/ray-project/kuberay/pull/1584#discussion_r1403618840)
* KubeRay currently supports Kubernetes 1.20 and above. However, it would be acceptable to only support versions 1.23 and higher, as the Kubernetes community has not supported version 1.23 since February 2023. (https://endoflife.date/kubernetes).
* I tried to test this PR with webhook disabled on Kubernetes 1.20.7. It works.

To conclude, it is fine for KubeRay to support v1.23+.

**Review by kevin85421 — APPROVED** [review]

**Comment by kevin85421** [comment]
Open an issue to track the progress: https://github.com/ray-project/kuberay/issues/1692

**Comment by davidxia** [comment]
thanks so much for merging, i'll take a look at that new issue when I get a chance.

I'll first make a doc PR to the [User Guides](https://docs.ray.io/en/latest/cluster/kubernetes/user-guides.html). I'm guessing you'd like a new page and item in that list for it?

**Comment by cermeng** [comment]
Good work! I have one question: it seems that helm chart installation is inconsistent with this pr and only `kustomize` installation for webhook is supported?

**Comment by davidxia** [comment]
@cermeng yes we didn't add that yet. But I think PRs are welcome.

**Pull request merged**
